File: C#/0001两数之和.cs
//暴力算法,双循环迭代，即可找到满足条件的数组下标，时间复杂度为O(n2)。

//优化算法，可以利用哈希表（C#用字典）在第一次迭代时先存储数组的数据，
//这样第二次迭代检索数据时可以利用索引快速查找，实现了用空间换时间。

public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        var dic = new Dictionary<int, int>();
        for(int i = 0; i < nums.Length; i++)
        {
            if (!dic.ContainsKey(nums[i]))
            {
                dic.Add(nums[i], i);
            }
        }
        for(int i = 0; i < nums.Length; i++)
        {
            int n = target - nums[i];
            if (dic.ContainsKey(n) && dic[n] != i)
            {
                return new int[2] { i, dic[n] };
            }
        }
        return null;
    }
}

File: C++/0001两数之和.cpp
//题目：给定一个整数数组 nums 和一个目标值 target，请你在该数组中找出和为目标值的那 两个 整数，并返回他们的数组下标。
//你可以假设每种输入只会对应一个答案。但是，数组中同一个元素不能使用两遍。


//更优算法，不必要先存储再查找，可以边存数据，边查找。

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map <int, int > map;
        vector<int> res;
        for (int i = 0; i < nums.size(); i++) {
            int n = target - nums[i];
            if (map.find(n) != map.end()) {
                res.push_back(map[n]);
                res.push_back(i);
                break;
            }
            else {
                map[nums[i]] = i;
            }
        }
        return res;
    }
};